pallets/amm/src/lib.rs

```diff
@@ -103,21 +115,201 @@ mod pallet {
+        ValueQuery,
+        GetDefault,
     >;
 
     /// A bag of liquidity composed by two different assets
     #[pallet::storage]
-    pub type Pools<T: Config<I>, I: 'static = ()> =
-        StorageMap<_, Blake2_128Concat, T::PoolId, Pool>;
+    #[pallet::getter(fn pools)]
+    pub type Pools<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+        _,
+        Twox64Concat,
+        CurrencyId,
+        Twox64Concat,
+        CurrencyId,
+        PoolLiquidityAmount,
+        ValueQuery,
+    >;
 
-    /// Auxiliary storage used to track pool ids
-    #[pallet::storage]
-    pub type PoolsCounter<T: Config<I>, I: 'static = ()> = StorageValue<_, T::PoolId, ValueQuery>;
+    #[pallet::call]
+    impl<T: Config<I>, I: 'static> Pallet<T, I> {
+        /// Allow users to add liquidity to a given pool
+        ///
+        /// - `pool`: Currency pool, in which liquidity will be added
+        /// - `liquidity_amounts`: Liquidity amounts to be added in pool
+        #[pallet::weight(10_000)]
+        #[transactional]
+        pub fn add_liquidity(
+            origin: OriginFor<T>,
+            pool: (CurrencyId, CurrencyId),
+            liquidity_amounts: (Balance, Balance),
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            let (base_asset, quote_asset) = Self::get_upper_currency(pool.0, pool.1);
+            let (base_amount, quote_amount) = liquidity_amounts;
+
+            if Pools::<T, I>::contains_key(base_asset, quote_asset) {
+                let pool_liquidity_amount: PoolLiquidityAmount =
+                    Self::pools(base_asset, quote_asset);
+                let ownership = sp_std::cmp::min(
+                    (base_amount.saturating_mul(pool_liquidity_amount.ownership))
+                        .checked_div(pool_liquidity_amount.base_amount)
+                        .expect("cannot overflow with positive divisor; qed"),
+                    (quote_amount.saturating_mul(pool_liquidity_amount.ownership))
+                        .checked_div(pool_liquidity_amount.quote_amount)
+                        .expect("cannot overflow with positive divisor; qed"),
+                );
+
+                Pools::<T, I>::try_mutate(
+                    &base_asset,
+                    &quote_asset,
+                    |pool_liquidity_amount| -> DispatchResult {
+                        pool_liquidity_amount.base_amount = pool_liquidity_amount
+                            .base_amount
+                            .checked_add(base_amount)
+                            .ok_or(ArithmeticError::Overflow)?;
+                        pool_liquidity_amount.quote_amount = pool_liquidity_amount
+                            .quote_amount
+                            .checked_add(quote_amount)
+                            .ok_or(ArithmeticError::Overflow)?;
+                        pool_liquidity_amount.ownership = ownership;
+                        Ok(())
+                    },
+                )?;
```

Comment:
`contains_key` + `Self::pools` + `try_mutate` will generate 3 reads and 1 write. We can optimize this via `try_mutate_exists`